### PR TITLE
feat(karma-webpack): only rerun tests for changed files

### PR DESCRIPTION
**Pull request by terencechow**
Currently a file change results in all tests being rerun. This change allows only the tests in the changed file to be rerun. If the changed file is not a test file then all tests will rerun.

This change is a copy of the work done in this fork https://github.com/aaronjensen/karma-webpack. However it is adapted to the current version of karma

**Please check if the PR fulfills these requirements**
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)
- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Other... Please describe:

**What is the current behavior?** (You can also link to an open issue here)

All tests are rerun when a file changes.

**What is the new behavior?**
Changed files are added to the global `__karmaWebpackManifest__` in watch mode. This means you can run only the tests found in the changed file, resulting in a much faster testing environment. If the changed file is not a test file then all tests are run. 

**Does this PR introduce a breaking change?**
- [ ] Yes
- [x] No

**Other information**:

This PR is a direct copy of the relevent snippets of code from https://github.com/aaronjensen/karma-webpack

### CODE REVIEW

**Comment by terencechow** [comment]
Note: I can also add the below snippet to the README.md if that is necessary:

```
// This gets replaced by karma webpack with the updated files on rebuild
var __karmaWebpackManifest__ = [];

// require all modules ending in "_test" from the
// current directory and all subdirectories
var testsContext = require.context(".", true, /_test$/);

function inManifest(path) {
  return __karmaWebpackManifest__.indexOf(path) >= 0;
}

var runnable = testsContext.keys().filter(inManifest);

// Run all tests if we didn't find any changes
if (!runnable.length) {
  runnable = testsContext.keys();
}

runnable.forEach(testsContext);
```

**Comment by d3viant0ne** [comment]
Before this goes anywhere, I want to talk to Aaron about this.

**Comment by aaronjensen** [comment]
@d3viant0ne thanks for checking in. I'm all for having this go in. Thank you for porting it @terencechow!

**Comment by terencechow** [comment]
FYI, I've been using this for a day now and there are some oddities. 

Namely there are times that the tests rerun the previous test again. So if I change `a.spec.js` it will run the tests in `a.spec.js`, but if I follow that by changing `b.spec.js` it will run tests in both a & b. Making a change in `b.spec.js` again will then only run the tests in `b.spec.js`. Almost as if it is keeping a running tab of the previous file changed.

I see the issue is the below line not filtering for just the last changed file. I'm not too familiar with webpack stats and if there is any details in the stats that indicate what the last file changed.

```
function isBuilt (module) { return module.rawRequest && module.built; }
var affectedFiles = stats.compilation.modules.filter(isBuilt)...
```

A `module` object in the `stats.compilation.modules` array is rather large so if anyone has any suggestions please feel free to mention them and I will update the PR

File: src/karma-webpack.js [hunk 1]
```diff
@@ -94,6 +96,36 @@ function Plugin(
...
+		function setTrue(acc, key) { acc[key] = true; return acc; }
+
+		var affectedFiles = stats.compilation.modules
+		  .filter(isBuilt)
+			.map(getId)
+			.reduce(setTrue, {})
+		var seen = {};
+
+		function findAffected(module) {
+			if (seen[module.rawRequest]) return;
```
Comment (michael-ciniawsky) [inline src/karma-webpack.js]: `\n` after the if block

File: src/karma-webpack.js [hunk 2]
```diff
@@ -94,6 +96,36 @@ function Plugin(
...
 
   compiler.plugin('done', function(stats) {
+		function isBuilt(module) { return module.rawRequest && module.built; }
+		function getId(module) { return module.rawRequest; }
+		function setTrue(acc, key) { acc[key] = true; return acc; }
+
+		var affectedFiles = stats.compilation.modules
+		  .filter(isBuilt)
+			.map(getId)
+			.reduce(setTrue, {})
```
Comment (michael-ciniawsky) [inline src/karma-webpack.js]: `\n` after the chain

File: src/karma-webpack.js [hunk 3]
```diff
@@ -94,6 +96,36 @@ function Plugin(
...
+		var seen = {};
+
+		function findAffected(module) {
+			if (seen[module.rawRequest]) return;
+			seen[module.rawRequest] = true;
+
+			if (affectedFiles[module.rawRequest]) return;
```
Comment (michael-ciniawsky) [inline src/karma-webpack.js]: Could this be joined into one branch ? 

```js
if (!module.rawRequest || !module.dependencies || affectedFiles[module.rawRequest]) return;
```

**Comment by fa-hmad** [comment]
Would love to have this go in. Any status?

**Comment by rynclark** [comment]
Love the idea. Is it possible to use webpack's dependency tree completely, so when editing a file that isn't a test file, it only runs the test files that depend on it?

We've refactored the code a whole bunch so if you don't want to continue with this I can take over, let me know

**Comment by michael-ciniawsky** [comment]
Closing as stalled and inactive for now, feel free to reopen :)

**Pull request closed without merging**